### DISCUSSION TITLE
How to lock versions of command line tools for iOS development

### DISCUSSION THREAD

**Discussion by ahmdyasser** [MobileNativeFoundation/discussions#220] [General]
There is a problem I couldn't figure and I'd love to hear any feedback or suggestions. 

When using tools like SwiftFormat, SwiftLint, or other tools that are not part of the project itself, they are just tools "around" the project. So, to me, it doesn't make any sense to add any of those tools as a dependency to my project through Swift Package Manger. Instead, I prefer to consume them through a package manger like Homebrew. 

However, this approach isn't that good, because Homebrew doesn't let you to lock the versions of your dependencies, and this is really problematic. It actually caused our CI to be blocked for some time because we don't have control over which version we are using.

There are package managers that allows you to use your own local version of those tools like Mise. In fact, Tuist migrated to it recently and suggesting to consume Tuist through Mise, which makes sense to me. But here another issue rises, which is, not all tools are registered in Mise registry.

Another solution is Mint, which seems promising because it "installs and runs Swift command line tool packages". But still, there are command line tools that are not Swift packages.

The issue that caused all CIs that are using Lokalise through Homebrew to break: https://github.com/lokalise/lokalise-cli-2-go/issues/152

Let me know if you want to discuss potential solutions to this issue

**Comment by mattrobmattrob** [comment]
I have used Bazel for this type of thing in a handful of projects, even if they aren't using Bazel to build the iOS component. It just requires the user to have Bazelisk installed then it's hermetic. You could build the tools from source and use cache or go the simpler route and reference the prebuilt binaries.

You can check the `BUILD.bazel` file in the root of this repo for an example that runs Swiftlint and a Bazel formatting tool, in this case: https://github.com/mattrobmattrob/bazel-ios-swiftui-template

**Comment by mattrobmattrob** [comment]
Similarly to my Bazel comment, I've also used `brew bundle` in the past for more controlled Homebrew "lockfile" control. It might fit the bill for you here: https://github.com/Homebrew/homebrew-bundle
